rust/plugin-lib/src/state_cache.rs

```diff
@@ -521,8 +521,10 @@ impl<'a, S: Default + Clone> PluginCtx<'a, S> {
     /// Updates the line cache to reflect this delta.
     fn update_line_cache(&mut self, delta: &RopeDelta) {
         let (iv, new_len) = delta.summary();
-        if let Some(n) = delta.as_simple_insert() {
-            assert_eq!(iv.size(), 0);
+        // A paste that replaces a selection is treated as a simple insert,
+        // but it screws up our logic, so we pass if the interval size > 0.
```

Comment:
can you comment specifically how it screws up the logic?

Reply:
Your understanding is correct: the strategy for updating the cache here is that we handle the two common editing cases (typing/deleting) efficiently, and then in the rest of the cases we just blow away the cache.

How this is getting confused is that we were assuming that a 'simple edit' meant an insertion without any deletions, but apparently an insertion that replaces a region was also getting treated as a simple edit. Our logic in `line_cache_simple_insert` is based on the idea that when updating line offsets further down in the cache, the only thing we need to keep track of is the length of the inserted text; but if there's also a deleted region that logic breaks.